### PR TITLE
Interest system feature

### PR DESCRIPTION
**Pull request by adsnaider**
This branch contains the interest system feature. Basically, users can add and remove their interests, and  then call a status update which will print all the new information about the interests the user had setup previously.

### CODE REVIEW

File: src/codeu/chat/client/commandline/Chat.java [hunk 1]
```diff
@@ -171,21 +188,31 @@ public void invoke(Scanner args) {
...
+    for (final UserContext user : context.allUsers()) {
+      if (user.user.name.equals(name)){
+        return user;
+      }
+    }
+    return null;
+  } 
+  private ConversationContext find(String title, UserContext user) {
+    for(final ConversationContext conversation : user.conversations()) {
+      if(title.equals(conversation.conversation.title)) {
```
Comment (alanhzhao) [inline src/codeu/chat/client/commandline/Chat.java]: Separate keywords like `if`, `else`, and `for` using spaces. Also add spaces between `)` and `{`.

There are a lot of occurrences spread across several files, so you might want to try a regex replace (or see if your IDE handles this).

File: src/codeu/chat/client/commandline/Chat.java [hunk 2]
```diff
@@ -276,14 +308,142 @@ public void invoke(Scanner args) {
...
+    // C-UNFOLLOW (unfollows an interest)
+    // 
+    // Adds a command that will allow the user to unfollow users and
+    // conversations to stop getting updates 
+    //
+    panel.register("i-remove", new Panel.Command(){
+      @Override
+      public void invoke(Scanner args){
+
+        ArrayList<String> arguments = new ArrayList<String>();
```
Comment (alanhzhao) [inline src/codeu/chat/client/commandline/Chat.java]: Change to`List<String> arguments = ...`. Use interface classes whenever possible.

File: src/codeu/chat/client/commandline/Chat.java [hunk 3]
```diff
@@ -276,14 +308,142 @@ public void invoke(Scanner args) {
...
+    // Adds a command that will report the status updates of the followed users
+    // and conversations
+    //
+    panel.register("status-update", new Panel.Command(){
+      @Override
+      public void invoke(Scanner args){
+
+        Collection<InterestStatus> allInterests = 
+            user.getController().statusUpdate(user);
+        if(allInterests != null && allInterests.size() >= 1){
```
Comment (alanhzhao) [inline src/codeu/chat/client/commandline/Chat.java]: Use `!allInterests.isEmpty()`

File: src/codeu/chat/server/Controller.java [hunk 4]
```diff
@@ -143,29 +148,113 @@ public ConversationHeader newConversation(Uuid id, String title, Uuid owner, Tim
...
+    Time lastUpdate = interest.lastUpdate;
+    InterestStatus result = null;
+    if (interest.type == Type.USER) {
+      User user = model.userById().first(id);
+      // Return all values with time higher than last Update
+      Iterable<ConversationHeader> headers = model.conversationByTime().
+        after(lastUpdate);
+      List<String> createdConversations = new ArrayList<>();
+      for (ConversationHeader header : headers) {
+        if (!header.owner.equals(user.id)) continue;
```
Comment (alanhzhao) [inline src/codeu/chat/server/Controller.java]: Don't need `continue.` Just don't negate the condition, and put the next line inside the `if` block.

File: src/codeu/chat/server/Controller.java [hunk 5]
```diff
@@ -143,29 +148,113 @@ public ConversationHeader newConversation(Uuid id, String title, Uuid owner, Tim
...
+  }
+
+  public void removeInterest(Uuid userId, Uuid interestId) {
+    model.removeInterest(userId, interestId);
+  }
+
+  // TODO(Adam): Create and send an ArrayList of InterestStatus objects instead of an array
+  // list of strings. Allow the client to figure out how to sort through the
+  // data.
+  public ArrayList<InterestStatus> interestStatus(Uuid user) {
```
Comment (alanhzhao) [inline src/codeu/chat/server/Controller.java]: Change this method signature, `userInterests`, and `result` to type  `List`.

File: src/codeu/chat/common/Interest.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    this.type = type;
+    lastUpdate = now;
+  }
+
+  @Override
+  public boolean equals(Object other) {
+    if (other == null) return false;
+    if (!(other instanceof Interest)) return false;
+    Interest that = (Interest) other;
+    if (!(this.id.equals(that.id))) return false;
```
Comment (alanhzhao) [inline src/codeu/chat/common/Interest.java]: Just do
`return this.id.equals(that.id) && this.interestId.equals(that.interestId) && ...`

**Review by alanhzhao — CHANGES_REQUESTED** [review]

File: src/codeu/chat/client/commandline/Chat.java (hunk 1 above)
Comment (adsnaider) [reply src/codeu/chat/client/commandline/Chat.java]: Done. I had to run a formatting script in all the code, so there are a lot of other changes regarding style as well. Sorry :/

File: src/codeu/chat/client/commandline/Chat.java (hunk 2 above)
Comment (adsnaider) [reply src/codeu/chat/client/commandline/Chat.java]: Done

File: src/codeu/chat/server/Controller.java (hunk 5 above)
Comment (adsnaider) [reply src/codeu/chat/server/Controller.java]: Done

File: src/codeu/chat/server/Controller.java (hunk 4 above)
Comment (adsnaider) [reply src/codeu/chat/server/Controller.java]: Done

File: src/codeu/chat/common/Interest.java (hunk 6 above)
Comment (adsnaider) [reply src/codeu/chat/common/Interest.java]: Done

File: src/codeu/chat/client/commandline/Chat.java (hunk 3 above)
Comment (adsnaider) [reply src/codeu/chat/client/commandline/Chat.java]: Done

File: src/codeu/chat/common/Interest.java [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+    if (!(other instanceof Interest))
+      return false;
+    Interest that = (Interest) other;
+    return this.id.equals(that.id) && this.interestId.equals(that.interestId)
+        && this.type.equals(that.type) && this.lastUpdate.equals(that.lastUpdate);
+  }
+
+  @Override
+  public int hashCode() {
+    int total = 3 * id.hashCode() + 7 * interestId.hashCode() + 13 * type.hashCode()
```
Comment (alanhzhao) [inline src/codeu/chat/common/Interest.java]: Use [`Objects.hash`](https://docs.oracle.com/javase/7/docs/api/java/util/Objects.html#hash(java.lang.Object...)).

File: src/codeu/chat/server/Controller.java [hunk 8]
```diff
@@ -143,29 +159,120 @@ public ConversationHeader newConversation(Uuid id, String title, Uuid owner, Tim
...
+  public List<InterestStatus> interestStatus(Uuid user) {
+    List<Uuid> userInterests = model.interests.get(user);
+    List<InterestStatus> result = new ArrayList<>();
+    Time now = Time.now();
+    if (userInterests == null)
+      return result;
+    for (Uuid interestId : userInterests) {
+      InterestStatus report = processInterest(interestId, now);
+      if (report == null)
+        continue;
```
Comment (alanhzhao) [inline src/codeu/chat/server/Controller.java]: Remove `continue`, negate `if` condition, move next line inside `if`

File: src/codeu/chat/server/Model.java [hunk 9]
```diff
@@ -67,6 +78,10 @@ public int compare(Time a, Time b) {
   private final Store<Time, Message> messageByTime = new Store<>(TIME_COMPARE);
   private final Store<String, Message> messageByText = new Store<>(STRING_COMPARE);
 
+  private final Store<Uuid, Interest> interestById = new Store<>(UUID_COMPARE);
+
+  public final HashMap<Uuid, ArrayList<Uuid>> interests = new HashMap<>();
```
Comment (alanhzhao) [inline src/codeu/chat/server/Model.java]: interests  should have type `Map`

File: src/codeu/chat/common/User.java [hunk 10]
```diff
@@ -59,4 +56,20 @@ public User(Uuid id, String name, Time creation) {
     this.creation = creation;
 
   }
+
+  @Override
+  public int hashCode() {
+    return 5 * id.hashCode() + 7 * name.hashCode() + 11 * creation.hashCode();
```
Comment (alanhzhao) [inline src/codeu/chat/common/User.java]: `Objects.hash`

**Review by alanhzhao — CHANGES_REQUESTED** [review]

File: src/codeu/chat/common/Interest.java (hunk 7 above)
Comment (adsnaider) [reply src/codeu/chat/common/Interest.java]: Done

File: src/codeu/chat/server/Controller.java (hunk 8 above)
Comment (adsnaider) [reply src/codeu/chat/server/Controller.java]: Done

File: src/codeu/chat/server/Model.java (hunk 9 above)
Comment (adsnaider) [reply src/codeu/chat/server/Model.java]: Done

File: src/codeu/chat/common/User.java (hunk 10 above)
Comment (adsnaider) [reply src/codeu/chat/common/User.java]: Done

**Review by alanhzhao — APPROVED** [review]

**Comment by alanhzhao** [comment]
There's more formatting polish to be done, but I'm okay with doing it later if you want.

As Github points out, you'll need to merge the latest changes master into this branch:

    git fetch origin
    git checkout -b interest-system-combined origin/interest-system-combined
    git merge master

**Pull request merged**